core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicatedMutationAction.java

```diff
@@ -0,0 +1,151 @@
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportResponse;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+/**
+ * Base class for transport actions that modify data in some shard like index, delete, and shardBulk.
+ */
+public abstract class TransportReplicatedMutationAction<
+            Request extends ReplicatedMutationRequest<Request>,
+            Response extends ReplicatedMutationResponse
+        > extends TransportReplicationAction<Request, WriteResult<Response>, Request, Response> {
+
+    protected TransportReplicatedMutationAction(Settings settings, String actionName, TransportService transportService,
+            ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, Supplier<Request> request,
+            String executor) {
+        super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
+                indexNameExpressionResolver, request, request, executor);
+    }
+
+    /**
+     * Called with a reference to the primary shard.
+     *
+     * @return the result of the write - basically just the response to send back and the translog location of the {@linkplain IndexShard}
+     *         after the write was completed
+     */
+    protected abstract WriteResult<Response> onPrimaryShard(IndexService indexService, IndexShard indexShard, Request request)
+            throws Exception;
+
+    /**
+     * Called once per replica with a reference to the {@linkplain IndexShard} to modify.
+     *
+     * @return the translog location of the {@linkplain IndexShard} after the write was completed
+     */
+    protected abstract Translog.Location onReplicaShard(Request request, IndexShard indexShard);
+
+    @Override
+    protected Tuple<Request, WriteResult<Response>> shardOperationOnPrimary(Request request) throws Exception {
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(request.shardId().id());
+        WriteResult<Response> result = onPrimaryShard(indexService, indexShard, request);
+        processAfterWrite(request.isRefresh(), indexShard, result.location);
```

Comment:
for the primary would it make sense to run `processAfterWrite(request.isRefresh(), indexShard, result.location);`
in the `asyncShardOperationOnPrimary` now since we are kinda waiting for the replicas and somebody else might have fsynced the tlog for us already?